Add Edit Enrolment command

dylkaw opened the pull request:
Closes #127, closes #22 , closes #21, closes #7 

What have you done in this PR?
* Added `EditEnrolmentCommand`, `EditEnrolmentCommandParser` and `EditEnrolmentDescriptor`
* Added the corresponding tests

dylkaw committed:
Create EnrolmentExistsPredicate

This predicate was created so that it is possible to check if an
Enrolment exists so that it can be edited.

dylkaw committed:
Create EditEnrolmentCommand

dylkaw committed:
Create EditEnrolmentCommandParser

dylkaw committed:
Add case for EditEnrolmentCommand

dylkaw committed:
Add setEnrolment method for Model

dylkaw committed:
Add error message for non-existing enrolment

dylkaw committed:
Allow editing of memberName and eventName

Currently we cannot edit the memberName and eventName of an Enrolment.

We should support this so editing name of Event or Member will edit
Enrolments.

Let’s add the memberName and eventName fields and their related methods
to EditEnrolmentDescriptor to facilitate this.

dylkaw committed:
Fix EditEnrolmentCommand equals method

Currently equals method returns true if hours and remark attribute are
equal.

This is incorrect since EditEnrolmentCommand contains memberName
and eventName as well.

Let’s update the equals method to compare the equality of
memberName and eventName with the other memberName and
eventName respectively.

dylkaw committed:
Create EditEnrolmentDescriptorBuilder

dylkaw committed:
Add setEnrolment method to test stubs

dylkaw committed:
Add EditEnrolmentDescriptor to CommandTestUtil

dylkaw committed:
Create EditEnrolmentDescriptorTest

dylkaw committed:
Add Enrolment typical indexes

dylkaw committed:
Add showEnrolmentAtIndex method

dylkaw committed:
Add EditEnrolmentCommandTest

dylkaw committed:
Create EditEnrolmentCommandParserTest

dylkaw committed:
Fix CheckStyle errors

dylkaw committed:
Merge branch 'master' into edit-enrolment

dylkaw committed:
Refactor Attendance to Enrolment

dylkaw committed:
Delete AttendanceTest

dylkaw committed:
Fix CheckStyle error in test

dylkaw committed:
Create EnrolmentExistsPredicateTest

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+        model.updateFilteredEnrolmentList(new EnrolmentExistsPredicate(enrolmentToCheck));
+        List<Enrolment> enrolmentToEditList = model.getFilteredEnrolmentList();
+        if (enrolmentToEditList.size() == 0) {
+            throw new CommandException(String.format(Messages.MESSAGE_ENROLMENT_DOES_NOT_EXIST,
+                    memberIndex.getOneBased(), eventIndex.getOneBased()));
```
Do you want to test this as warned by codecov?

dylkaw replied:
Done!

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+        List<Member> lastShownMemberList = model.getFilteredMemberList();
+        List<Event> lastShownEventList = model.getFilteredEventList();
+
+        if (memberIndex.getZeroBased() >= lastShownMemberList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_MEMBER_DISPLAYED_INDEX);
+        }
+        if (eventIndex.getZeroBased() >= lastShownEventList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_EVENT_DISPLAYED_INDEX);
+        }
```
In #179 we have a new MESSAGE to show if both the event and member displayed are invalid, I think we should follow that implementation as user might be told only their member index is invalid when it's both.

dylkaw replied:
Ok, we can do that once #179 is merged.

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+        Member targetMember = lastShownMemberList.get(memberIndex.getZeroBased());
+        Event targetEvent = lastShownEventList.get(eventIndex.getZeroBased());
+        Enrolment enrolmentToCheck =
+                new Enrolment(targetMember.getName(), targetEvent.getName(), new Hours("0"), new Remark("HOLDER"));
```
Should we rename the Remark to something like "UNUSED REMARK" to make it more obvious that it's unused since HOLDER might be confusing

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+            throw new CommandException(String.format(Messages.MESSAGE_ENROLMENT_DOES_NOT_EXIST,
+                    memberIndex.getOneBased(), eventIndex.getOneBased()));
+        }
+        assert enrolmentToEditList.size() == 1 : "There should not be duplicate enrolments";
```
Good use of defensive programming!

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+     */
+    private static Enrolment createEditedEnrolment(Enrolment enrolmentToEdit,
+                                                    EditEnrolmentDescriptor editEnrolmentDescriptor) {
+        assert enrolmentToEdit != null;
```
Why don't we use RequireNonNull here? We could ensure that editEnrolmentDescriptor is not null too

dylkaw replied:
Good point, I've changed it.

wr1159 commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+        private Hours hours;
+        private Remark remark;
+
+        public EditEnrolmentDescriptor() {}
```
Do we need this empty constructor if we have the other constructor?

ph-nathan replied:
I think yes cause in line 38 of EditEnrolmentCommandParser file, there is a use for this constructor

wr1159 commented on src/main/java/seedu/ccacommander/logic/parser/EditEnrolmentCommandParser.java:
```diff
@@ -0,0 +1,65 @@
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
```
Oh actually I realised that we have arePrefixesPresent in other files which seem to have the same implementation, I think we can abstract this out to it's own parseUtil file in the future.

dylkaw replied:
I agree, this is something we can consider in the future.

wr1159 commented on src/test/java/seedu/ccacommander/logic/commands/CommandTestUtil.java:
```diff
@@ -144,6 +147,23 @@ public class CommandTestUtil {
                 .withDate(VALID_DATE_BOXING)
                 .withTags(VALID_TAG_BOXING).build();
     }
+
+    public static final EditEnrolmentCommand.EditEnrolmentDescriptor DESC_AMY_AURORA;
+    public static final EditEnrolmentCommand.EditEnrolmentDescriptor DESC_BOB_BOXING;
+
+    static {
+        DESC_AMY_AURORA = new EditEnrolmentDescriptorBuilder()
+                .withMemberName(VALID_NAME_AMY)
+                .withEventName(VALID_NAME_AURORA)
+                .withHours(VALID_HOURS_AURORA)
+                .withRemark(VALID_REMARK_AURORA).build();
+        DESC_BOB_BOXING = new EditEnrolmentDescriptorBuilder()
+                .withMemberName(VALID_NAME_BOB)
+                .withEventName(VALID_NAME_BOXING)
+                .withHours(VALID_HOURS_BOXING)
+                .withRemark(VALID_REMARK_BOXING).build();
+    }
+
```
Why are we using this method of instantiating these constants?

dylkaw replied:
Given that we have a method to generate these constants, it is easier to use and more readable than just concatenating String literals.

wr1159 reviewed (changes requested):
LGTM, just a few comments for you to look through

ph-nathan commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+                                                    EditEnrolmentDescriptor editEnrolmentDescriptor) {
+        assert enrolmentToEdit != null;
+
+        Name memberName = editEnrolmentDescriptor.getMemberName().orElse(enrolmentToEdit.getMemberName());
+        Name eventName = editEnrolmentDescriptor.getEventName().orElse(enrolmentToEdit.getEventName());
```
small nit but since these two fields are unmodifiable through editEnrolment, I think can just put enrolmentToEdit.getMemberName() for conciseness :0 Its not too significant though

ph-nathan commented on src/main/java/seedu/ccacommander/logic/commands/EditEnrolmentCommand.java:
```diff
@@ -0,0 +1,236 @@
+     * corresponding field value of the event.
+     */
+    public static class EditEnrolmentDescriptor {
+        private Name memberName;
+        private Name eventName;
```
should we leave Names related fields and methods in EditEnrolmentDescriptor out, cause i checked and it seemed they arent used in any place below. And because memberName and eventName are not supposed to be updated. But either way it wouldnt really affect 👍

dylkaw replied:
I think they will be required for editing memberName of a member, or eventName of an event, as these fields will then have to be edited. For example, editing a member's Name will result in the editing of all their Enrolments.

ph-nathan reviewed (approved):
LGTM, left a few comments for discussion but even if don't change them I think it would still be fine. Grate job :0

dylkaw committed:
Add enrolmentDoesNotExist test

dylkaw committed:
Add quality of life changes

Some minor changes:
- Changed remark "HOLDER" to "UNUSED REMARK"
- Added requireNonNull for enrolmenttoEdit
- Added requireNonNull for editEnrolmentDescriptor

wr1159 reviewed (approved):
LGTM
